### PR TITLE
R4R: Finish the new gov

### PR DESCRIPTION
**Pull request by MrXJC**

### CODE REVIEW

File: modules/gov/keeper.go [hunk 1]
```diff
@@ -573,4 +582,130 @@ func (keeper Keeper) SetTerminatorPeriod(ctx sdk.Context, height int64) {
...
+
+
+func (keeper Keeper) IsMoreThanMaxProposal(ctx sdk.Context, pl ProposalLevel) (uint64,bool) {
+	switch pl {
+	case ProposalLevelCritical:
+		return keeper.GetCriticalProposalNum(ctx), keeper.GetCriticalProposalNum(ctx) >= 1
+	case ProposalLevelImportant:
+		return keeper.GetImportantProposalNum(ctx),keeper.GetImportantProposalNum(ctx) >= 3
+	case ProposalLevelNormal:
+		return keeper.GetNormalProposalNum(ctx), keeper.GetNormalProposalNum(ctx) >= 5
```
Comment (wukongcheng) [inline modules/gov/keeper.go]: The `MaxProposal` should be a gov param instead of magic number

File: modules/gov/keeper.go [hunk 2]
```diff
@@ -573,4 +582,130 @@ func (keeper Keeper) SetTerminatorPeriod(ctx sdk.Context, height int64) {
...
+	default:
+		panic("There is no level for this proposal")
+	}
+}
+
+func (keeper Keeper) BeginProposal(ctx sdk.Context, p govtypes.Proposal) {
```
Comment (wukongcheng) [inline modules/gov/keeper.go]: BeginProposal  -> AddProposalNum

File: modules/gov/keeper.go [hunk 3]
```diff
@@ -573,4 +582,130 @@ func (keeper Keeper) SetTerminatorPeriod(ctx sdk.Context, height int64) {
...
+	case ProposalLevelImportant:
+		keeper.AddImportantProposalNum(ctx)
+	case ProposalLevelNormal:
+		keeper.AddNormalProposalNum(ctx)
+	default:
+		panic("There is no level for this proposal which type is " + p.GetProposalType().String())
+	}
+}
+
+func (keeper Keeper) EndProposal(ctx sdk.Context, p govtypes.Proposal) {
```
Comment (wukongcheng) [inline modules/gov/keeper.go]: EndProposal -> SubProposalNum

File: modules/gov/keeper_keys.go [hunk 4]
```diff
@@ -82,4 +82,7 @@ func KeyInactiveProposalQueueProposal(endTime time.Time, proposalID uint64) []by
 var (
 	KeyTerminatorHeight = []byte("TerminatorHeight")
 	KeyTerminatorPeriod = []byte("TerminatorPeriod")
+	KeyCriticalProposal = []byte("CriticalProposalList")
```
Comment (wukongcheng) [inline modules/gov/keeper_keys.go]: wrong name 'CriticalProposalList'

File: modules/gov/handler.go [hunk 5]
```diff
@@ -259,19 +277,48 @@ func EndBlocker(ctx sdk.Context, keeper Keeper) (resTags sdk.Tags) {
...
+			keeper.DeleteDeposits(ctx, activeProposal.GetProposalID())
+			activeProposal.SetStatus(govtypes.StatusRejected)
+			action = tags.ActionProposalRejected
+		} else {
+			return resTags
+		}
 		keeper.RemoveFromActiveProposalQueue(ctx, activeProposal.GetVotingEndTime(), activeProposal.GetProposalID())
+		activeProposal.SetTallyResult(tallyResults)
+		activeProposal.SetVotingEndTime(ctx.BlockHeader().Time)
+		keeper.SetProposal(ctx, activeProposal)
```
Comment (wukongcheng) [inline modules/gov/handler.go]: put these tally and result processing codes into a function `ProcessTallyResult`

File: modules/gov/handler.go [hunk 6]
```diff
@@ -286,13 +286,25 @@ func EndBlocker(ctx sdk.Context, keeper Keeper) (resTags sdk.Tags) {
 		resTags = resTags.AppendTag(tags.Action, action)
 		resTags = resTags.AppendTag(tags.ProposalID, []byte(string(proposalID)))
 
+        for _, valAddr := range keeper.GetValidatorSet(ctx,proposalID){
+        	if _,ok := votingVals[valAddr.String()]; !ok {
+				val := keeper.ds.GetValidatorSet().Validator(ctx, valAddr)
```
Comment (wukongcheng) [inline modules/gov/handler.go]: if val is nil ?

Comment (wukongcheng) [reply modules/gov/handler.go]: This issue will panic the whole blockchain

File: modules/gov/handler.go [hunk 7]
```diff
@@ -286,13 +286,25 @@ func EndBlocker(ctx sdk.Context, keeper Keeper) (resTags sdk.Tags) {
...
+				keeper.ds.GetValidatorSet().Slash(ctx,
+					val.GetConsAddr(),
+					ctx.BlockHeight(),
+					val.GetPower().RoundInt64(),
+					GetTallyingProcedure(ctx).GovernancePenalty)
+			}
+		}
+
 		keeper.EndProposal(ctx, activeProposal)
+		keeper.DeleteValidatorSet(ctx,activeProposal.GetProposalID())
```
Comment (wukongcheng) [inline modules/gov/handler.go]: Where to DeleteValidatorSet for the CriticalProposal ?

File: modules/gov/handler.go [hunk 8]
```diff
@@ -259,19 +277,64 @@ func EndBlocker(ctx sdk.Context, keeper Keeper) (resTags sdk.Tags) {
...
+						val.GetPower().RoundInt64(),
+						GetTallyingProcedure(ctx).GovernancePenalty)
+				}
+			}
+		}
+
+		keeper.SubProposalNum(ctx, activeProposal)
+		keeper.DeleteValidatorSet(ctx, activeProposal.GetProposalID())
+	}
+	activeIterator.Close()
```
Comment (HaoyangLiu) [inline modules/gov/handler.go]: defer activeIterator.Close()

**Pull request merged**